src/coreclr/jit/fgprofile.cpp

```diff
@@ -439,443 +447,1176 @@ void BlockCountInstrumentor::InstrumentMethodEntry(Schema& schema, BYTE* profile
-    //
-    int classSchemaIndex = block->bbClassSchemaIndex;
-    assert((classSchemaIndex >= 0) && (classSchemaIndex < (int)schema.size()));
-
-    ClassProbeInserter                    insertProbes(schema, profileMemory, &classSchemaIndex, m_instrCount);
-    ClassProbeVisitor<ClassProbeInserter> visitor(m_comp, insertProbes);
-    for (Statement* stmt : block->Statements())
-    {
-        visitor.WalkTree(stmt->GetRootNodePointer(), nullptr);
-    }
+    JITDUMP("\nEfficientEdgeCountInstrumentor: preparing for instrumentation\n");
+    m_comp->WalkSpanningTree(this);
+    JITDUMP("%u blocks, %u probes (%u on critical edges)\n", m_blockCount, m_probeCount, m_edgeProbeCount);
 }
 
 //------------------------------------------------------------------------
-// ClassProbeInstrumentor::SuppressProbes: clean up if we're not instrumenting
+// EfficientEdgeCountInstrumentor:BuildSchemaElements: create schema
+//   elements for the probes
 //
-// Notes:
-//   Currently we're hijacking the gtCallStubAddr of the call node to hold
-//   a pointer to the profile candidate info.
+// Arguments:
+//   block -- block to instrument
+//   schema -- schema that we're building
 //
-//   We must undo this, if not instrumenting.
+// Todo: if required to have special entry probe, we must also
+//  instrument method entry with a block count.
 //
-void ClassProbeInstrumentor::SuppressProbes()
+void EfficientEdgeCountInstrumentor::BuildSchemaElements(BasicBlock* block, Schema& schema)
 {
-    unsigned                                 cleanupCount = 0;
-    SuppressProbesFunctor                    suppressProbes(cleanupCount);
-    ClassProbeVisitor<SuppressProbesFunctor> visitor(m_comp, suppressProbes);
-
-    for (BasicBlock* block = m_comp->fgFirstBB; (block != nullptr); block = block->bbNext)
+    // Walk the bbProbeList, emitting one schema element per...
+    //
+    for (Probe* probe = (Probe*)block->bbProbeList; probe != nullptr; probe = probe->next)
     {
-        if ((block->bbFlags & BBF_HAS_CLASS_PROFILE) == 0)
+        // Probe is for the edge from block->target.
+        //
+        BasicBlock* const target = probe->target;
+
+        // Remember the schema index for this probe
+        //
+        assert(probe->schemaIndex == -1);
+        probe->schemaIndex = (int)schema.size();
+
+        // Assign the current block's IL offset into the profile data.
+        // Use the "other" field to hold the target block IL offset.
+        //
+        int32_t sourceOffset = (int32_t)block->bbCodeOffs;
+        int32_t targetOffset = (int32_t)target->bbCodeOffs;
+
+        // We may see empty BBJ_NONE BBF_INTERNAL blocks that were added
+        // by fgNormalizeEH.
+        //
+        // We'll use their bbNum in place of IL offset, and set
+        // a high bit as a "flag"
+        //
+        if ((block->bbFlags & BBF_INTERNAL) == BBF_INTERNAL)
         {
-            continue;
+            sourceOffset = block->bbNum | IL_OFFSETX_CALLINSTRUCTIONBIT;
```

Comment:
Seems like you need to define your own high bit here, and maybe even accessor/settors, for extreme clarity.